### ISSUE TITLE
Cannot Parse a GM SoundFont

### ISSUE THREAD

**Issue by billtj** [notator/SimpleSoundFontSynthHost#1]
I changed the soundfont from "Arachno1.0selection-grand piano.sf2" to "TimGM6mb.sf2", which is a full GM soundfont.  However, the program only finds patch 60 (French Horns) to patch 127 (Gun Shot) in bank 0; in other words, patches 0 - 59 are missing.  Probably there are some errors in the parser?

**Comment by notator** [comment]
Thanks very much for testing with another font, and for the feedback! :-)

That's very strange. The parser works fine for the other Arachno selections I used in
[WebMIDISynthHost](https://github.com/notator/WebMIDISynthHost). I suspect that you also wrote the following to the [google thread](https://code.google.com/p/chromium/issues/detail?id=503270):

>Hi, I took a look at SimpleSoundFontSynthHost, and it appears promising. However, when I changed the soundfont from "Arachno1.0selection-grand piano.sf2" (6767 KB) to "TimGM6mb.sf2" (5854 KB) there are some errors. Since the two soundfonts are about the same size, and "TimGM6mb.sf2" is full GM, can we modify the codes so that it will work with "TimGM6mb.sf2"?

Are you sure there are no errors in TimGM6mb.sf2? If its a full GM soundfont I'd expect it to be *much* bigger than Arachno1.0selection-grand piano.sf2, which only contains a grand piano.

Where can I download TimGM6mb.sf2? I'm not really an expert with soundfonts either (We really need someone like that on board!), but I'll take a closer look when I've got a copy.

 >(I haven't grasped yet the structure of soundfont in JavaScript; I managed only to get Bird sound (GM #123) to sound when I set the bankIndex to 128.)

Bank 128 contains "percussion". If you set bank 128, you get a different sound per note (one of them probably being a bird tweet). You could try again, and try different notes. :-)

>If this works with a GM soundfont, then this method is as good as (or even better than) using some plugins.

*When* it works it will be better! :-)

**Comment by billtj** [comment]
I got "TimGM6mb.sf2" from #25 of the Chromium discussion (https://code.google.com/p/chromium/issues/detail?id=503270#c25). If you follow the instructions there, you will be at VirtualMIDISynth website, and it has several GM soundfont links, with "TimGM6mb.sf2" as the smallest (5,854 KB) and "CrisisGeneralMidi 3.01" as the largest (1,649,602 KB) (!).

I got Viena soundfont editor from http://www.synthfont.com/; therefore I can see that "TimGM6mb.sf2" has 128 GM sounds and 8 drum kits. The JavaScript program also found that the soundfont contains 136 presets, but after parsing, it did not get all the 136 presets.

**Comment by billtj** [comment]
Hi, actually by following your pointer to "gree sf2synth.js", it seems that sf2synth.js is using the soundfont "A320U.sf2" (9,472 KB).  Probably we should concentrate on this soundfont because it has been used successfully by sf2synth.js.  Also, this soundfont contains additional sounds in addition to GM set, with a total of 325 presets!

**Comment by notator** [comment]
I think we need to test with several different fonts. The more the better. But I'm currently up to my ears working on [Chris Wilson's synth](https://lists.w3.org/Archives/Public/public-audio-dev/2015Dec/0001.html). Also, I've been given a hot tip about how to solve [issue #7 in the parent project](https://github.com/notator/WebMIDISynthHost/issues/7), and want to solve that before moving on to other fonts. Every bug solved teaches one more about the way the whole thing works, so I'd rather dig deep than go wide.
I'll get to your problem eventually, of course!

**Comment by billtj** [comment]
Hi James, after investigating further, I think we can use "gree/sf2synth.js" as simple as this:

In the HTML file, include

    <script src="sf2synth.min.js"></script>

In the JavaScript file, include

    var url = './A320U.sf2';
    var wml = new SoundFont.WebMidiLink();
    wml.setLoadCallback(function(arraybuffer) {
    });
    wml.setup(url);

as stated in sf2synth.js github page.

Then, for example, to have "synth.noteOn(0, 64, 100);", you may write

    var event = document.createEvent('Event');
    event.initEvent('midi');
    event.data = 'midi,' + (0x90).toString(16) + ',' + (64).toString(16) + ',' + (100).toString(16);
    wml.ha(event);

or something similar. "ha" is an alias for "onmessage" in the sf2synth.min.js file.

Now, the problem is timing, because in Web MIDI API, it provides

    void send (sequence<octet> data, optional DOMHighResTimeStamp timestamp);

I really need to use the timestamp.  Do you know the equivalent of the above in Web Audio API?  Or have you looked at "mudcube/MIDI.js"?  Now I will start examining MIDI.js to see whether and how the timing is done.

**Comment by billtj** [comment]
Hi James, after browsing "mudcube/MIDI.js", it seems that what are needed are

    MIDI.noteOn(channel, note, velocity, delay);
    MIDI.noteOff(channel, note, delay);

The problem with MIDI.js is that it appears that it uses "soundfonts" of .mp3 files, not the .sf2 files. Because I need to use .sf2 files and the delay, so it will be either to modify MIDI.js to work with .sf2 files, or to modify sf2synth.js to provide the delay attribute. I guess the latter is easier.  So I will try to access the Web Audio API's AudioBufferSourceNode's start method:

    void start (optional double when = 0
              , optional double offset = 0
              , optional double duration);

in sf2synth.js.  Who knows, maybe it is already provided somewhere inside the codes? :)

**Comment by notator** [comment]
I gave up on MIDI.js for the same reason: I want to use proper sf2 files.

[Web MIDI API §4.6](http://webaudio.github.io/web-midi-api/#methods-2)  says that the timestamp argument to send() is optional. If you set it to 0, the sound plays immediately. Maybe you can organize the timing yourself.

You could also look at gree's [smfPlayer](https://github.com/gree/smfplayer.js), to see how they do timing.
Ryoya Kawai's [smfplayer](https://github.com/ryoyakawai/smfplayer) uses the gree smfplayer, and his code might be easier to understand.
The problem for me, in both cases, is that they don't allow the host program to have an independent GUI. That's where this project started.

Best of luck!
James

**Comment by billtj** [comment]
Hi James, thank you for the information; I will take a look at them. I just have some questions:

1) Actually if we look at WebMidiLink Specification (http://www.g200kg.com/en/docs/webmidilink/spec.html) it contains similar (or actually better) info on how to interface with sf2synth.js. In my opinion, the host program can have an independent GUI. Therefore, is including sf2synth.min.js too much? (In other words, do you want something smaller than sf2synth.min.js?)

2) We both agree that the drawback of MIDI.js is the lack of support for .sf2 files. Have you taken a look at "gleitz/midi-js-soundfonts"? (It only provides one soundfont, FluidR3_GM.sf2, and I don't know how easy it is to get other .sf2 files to work.)

3) One of the excellent things about MIDI.js is its "perfect timing", and I have noticed this when I use the timestamp to send() in Web MIDI API. But dealing with MIDI input in real-time is different story: I have an external device (MIDI Sport Uno) hooked to my PC, and when I followed the example in Web MIDI API, there is always a delay of hundreds of milliseconds between when I push a key on the musical keyboard and when Chrome responds. I know that there is almost no delay when we interact using the PC mouse. Do you have any experience of dealing with external input such as from an external musical keyboard?  (I don't know which causes the long delay, whether the MIDI Sport Uno or the PC MIDI system, because when I hook the MIDI Sport Uno directly to the plugin VirtualMIDISynth, the delay is the same; so it has nothing to do with Web MIDI API.)

Best regards,
Bill

**Comment by notator** [comment]
Hi Bill,

1) WebMIDIlink uses strings to send messages. Long term, that means too much time converting arguments to bytes. Its like old-fashioned scripting: okay for simple cases, but its not the long term answer. The real answer is at the level of MIDI itself. I'm afraid there's no way round programming MIDI at the bits and bytes level.

2) Took a look at gleitz/midi-js-soundfonts. If understand it correctly, its a program for converting sf2 files to folders of mp3s, so that MIDI.js can use them. (We decided yesterday that we want to use real sf2s.) I haven't looked *very* closely at MIDI.js, but I don't think its trying to solve the same problem as I am. Its really just about play-back. I'm well on the way to solving a more general problem, and have the feeling I don't want to look back.

3) 
> dealing with MIDI input in real-time is different story

Exactly.
> Do you have any experience of dealing with external input such as from an external musical keyboard?

I used the VirtualMIDISynth for a couple of years without having to worry about latency. Indeed, I started using it because the Microsoft GS Wavetable Synth had such terrible latency problems.

You could try my [SimpleMIDISynthHost2](http://james-ingram-act-two.de/open-source/SimpleMIDISynthHost2/host.html). Its expecting input from a keyboard, and any knobs on the keyboard that send control messages. It plays Chris Wilson's synth fine from my keyboard. I think the latency is probably okay, though the default sound has a rather spongy attack, so its a bit difficult to be sure. If you have a keyboard that can send control messages (knobs) you could try to change the attack to get a better idea.

I think I've solved the general problem for Web Audio synthesizers. As always, if something like this is going to take off, it needs to be of interest to a wider community. There are actually quite a few people who have created synthesizers like Chris's, and they can all use the same strategy to make their devices accessible to any kind of Web MIDI program (with or without live input). The synths can even be made to play standard MIDI files... I can also imagine synths being created to produce much simpler sounds than Chris's. Bells or modulated flutes should be no problem, for example.

Apart from synthesizer authors/users, there is also a large and experienced community working with sf2 files. And I assume that there are *lots* of people who want to use grand pianos and harpsichords etc. Using an sf2 editor, web authors can even decide for themselves which instruments/sounds they want playing on their sites. Maybe we only need *one* sf2 player, for the whole sf2 community to get interested. Its a case of either waiting for me to dig in until I've got the thing working properly, or maybe some expert will get wind of this project and lend me a hand. I've already got a few leads, but professionals are always very busy... We'll see... :-)

best,
James

**Comment by billtj** [comment]
Hi James,

I guess my objective is different from yours. When Web MIDI API came out, I just wanted the PC to be able to generate some "semi-real-time" sounds using the built-in Microsoft GS Synth, and people don't have to install anything (of course, except the Chrome browser) on Windows. By "semi-real-time", I mean it is more interactive than just playing Standard MIDI Files, but not to the level of requiring an external musical keyboard. Therefore, when Chrome suddenly did not allow Microsoft GS Synth anymore, I kind of left the project behind, until somebody pointed us to use VirtualMIDISynth. It is better than using Microsoft GS Synth in the sense that people can choose their own soundfonts, but I guess some people don't, or can't, install VirtualMIDISynth, such as in the public PCs in the libraries. Then things got even better when you posted the message that your system does not require any plugins. So for me, after I finish modifying the code in sf2synth.js to add the delay, I am all set, because initially I just want to use Chrome with Microsoft GS Synth only.

Now, if you really want to do some MIDI (or Audio) programming at the bits and bytes level, you may consider Google Native Client. I think there will be some limitations, because you basically transform some JavaScript codes to either C or C++ codes. In other words, things that you can do in C or C++ are constrained by the Native Client environment. But if you don't want to use strings to interface with sf2synth, I think you can send the real bits and bytes by using Native Client.

I have one more question to you; how, or through what device, do you connect your musical keyboard (with all the knobs) to your PC?

Best regards,
Bill

**Comment by notator** [comment]
Hi Bill,

>So for me, after I finish modifying the code in sf2synth.js to add the delay, I am all set, because initially I just want to use Chrome with Microsoft GS Synth only.

I'll be starting serious work on my (originally gree's) soundFont synth tomorrow. That includes looking at your "TimGM6mb.sf2" problem. Note that this approach means that serious applications can be designed even without assuming that there's a MIDI keyboard plugged in. The host can program the GUI just using the generally available input devices (mouse, QWERTY keyboard or touch screen). See the first [SimpleSoundFontSynthHost](http://james-ingram-act-two.de/open-source/SimpleSoundFontSynthHost/host.html) or [WebMIDISynthHost](http://james-ingram-act-two.de/open-source/WebMIDISynthHost/host.html). For the latter, try setting different synths...

>Now, if you really want to do some MIDI (or Audio) programming at the bits and bytes level, you may consider Google Native Client. [...] I think you can send the real bits and bytes by using Native Client.

I didn't know Google Native Client, so took a look (thanks for the tip), but using it would land me with a few other issues. The bits and bytes can be programmed directly in Javascript, so I may as well stay with that. Also, Native Client seems to be a proprietary system. Wikipedia quotes Mozilla's vice president of products, Jay Sullivan, who said that:
>.. Mozilla has no plans to run native code inside the browser, as "These native apps are just little black boxes in a webpage. [...] We really believe in HTML, and this is where we want to focus.

I too want to focus on the lowest level of browser programming (Javascript), so that I'm immune from events such as Google dropping Native Client sometime. There is, on the other hand, a big effort being made to imrove Javascript. I think its extremely unlikely to be dropped in the next couple of decades.

>I have one more question to you; how, or through what device, do you connect your musical keyboard (with all the knobs) to your PC?

I've got an [E-MU Xboard 49](http://www.creative.com/emu/products/product.aspx?category=532&pid=13558). It has two connections to my computer: 1) ordinary USB, and 2) its MIDI output is attached to a [TS22 PCI](http://www.amazon.de/Terrasoniq-TS22-PCI-Recording-Interface/dp/B00178PQ62) sound card.

best,
James

**Comment by billtj** [comment]
Hi James,

Thanks for all the pointers.  So you got a musical keyboard (Xboard 49) with a built-in USB port; I think I will do the same thing.

Regarding sf2synth.js, when I tried it with both soundfonts "A320U.sf2" and "TimGM6mb.sf2", I got a slightly different problem that your issues #6, #7, and #8.  When I send a "note off", sometimes (and it happens usually the first time), there is an attack with the same velocity as in the previous "note on" value; in other words, it creates a "spike" before the sound is cut off.  Maybe I will take a look at what you have done in your codes after I finished adding the delay.

Best regards,
Bill

**Comment by notator** [comment]
Hi Bill,

Its great to be able to compare notes about issues in sf2synth.js! Its certainly not bug free, and all information helps. There's nobody answering at the gree site, so I think we should track those issues here.
I've created a new issue for your new noteOff observation, and am moving issues #6, #7 and #8 from  WebMIDISynthHost repository here as well.

By the way: looking at the E-MU Xboard 49 link I gave you, they seem to have upgraded the specification. Keys now send aftertouch. Keys on *my* E-MU Xboard 49 send channel pressure. Its  basically the same piece of equipment, but a couple of years old now. I'm very happy with it.
